kubevirt_handler/datadog_checks/kubevirt_handler/check.py

```diff
@@ -0,0 +1,119 @@
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from typing import Any  # noqa: F401
+
+from datadog_checks.base import OpenMetricsBaseCheckV2, is_affirmative
+from datadog_checks.base.checks.openmetrics.v2.transform import get_native_dynamic_transformer
+
+from .metrics import METRICS_MAP
+
+
+class KubeVirtHandlerCheck(OpenMetricsBaseCheckV2):
+    # This will be the prefix of every metric and service check the integration sends
+    __NAMESPACE__ = "kubevirt_handler"
+    DEFAULT_METRIC_LIMIT = 0
+
+    def __init__(self, name, init_config, instances):
+        super(KubeVirtHandlerCheck, self).__init__(name, init_config, instances)
+        self.check_initializations.appendleft(self._parse_config)
+        self.check_initializations.append(self._init_base_tags)
+        self.check_initializations.append(self._configure_additional_transformers)
+
+    def check(self, _):
+        # type: (Any) -> None
+
+        if self.kubevirt_handler_healthz_endpoint:
+            self._report_health_check(self.kubevirt_handler_healthz_endpoint)
+        else:
+            self.log.warning(
+                "Skipping health check. Please provide a `kubevirt_handler_healthz_endpoint` to ensure the health of the KubeVirt Handler."  # noqa: E501
+            )
+
+        super().check(_)
+
+    def _report_health_check(self, health_endpoint):
+        try:
+            self.log.debug("Checking health status at %s", health_endpoint)
+            response = self.http.get(health_endpoint, verify=self.tls_verify)
+            response.raise_for_status()
+            self.gauge("can_connect", 1, tags=[f"endpoint:{health_endpoint}", *self.base_tags])
+        except Exception as e:
+            self.log.error(
+                "Cannot connect to KubeVirt Handler HTTP endpoint '%s': %s.\n",
+                health_endpoint,
+                str(e),
+            )
+            self.gauge("can_connect", 0, tags=[f"endpoint:{health_endpoint}", *self.base_tags])
+            raise
+
+    def _parse_config(self):
+        self.kubevirt_handler_healthz_endpoint = self.instance.get("kubevirt_handler_healthz_endpoint")
+        self.kubevirt_handler_metrics_endpoint = self.instance.get("kubevirt_handler_metrics_endpoint")
+        self.kube_namespace = self.instance.get("kube_namespace")
+        self.pod_name = self.instance.get("kube_pod_name")
+        self.tls_verify = is_affirmative(self.instance.get("tls_verify"))
+
+        self.scraper_configs = []
+
+        instance = {
+            "openmetrics_endpoint": self.kubevirt_handler_metrics_endpoint,
+            "namespace": self.__NAMESPACE__,
+            "enable_health_service_check": False,
+            "tls_verify": self.tls_verify,
+        }
+
+        self.scraper_configs.append(instance)
+
+    def _init_base_tags(self):
+        self.base_tags = [
+            "pod_name:{}".format(self.pod_name),
+            "kube_namespace:{}".format(self.kube_namespace),
+        ]
```

Comment:
Why would you specifically need these?

Reply:
@steveny91 I'm not sure if the agent will attach these tags automatically since it's running on K8s, but the `agent status` output only shows these host tags:
```yaml
host tags:
    cluster_name:XXXXX
    kube_cluster_name:XXXXX
    kube_node:XXXXX
```
And having these tags provides better tracking for the kubevirt resources, especially the pod name since you could have multiple pods behind the same kubevirt service.

Reply:
These wouldn't show in the host level tags. But they will show in the metrics/instance level tags. I.e:

```
agent configcheck

...
=== istio check ===
Configuration provider: file
Configuration source: file:/etc/datadog-agent/conf.d/istio.d/auto_conf.yaml
Config for instance ID: istio:60f581df689cb765
istio_mesh_endpoint: http://10.108.0.9:15020/stats/prometheus
send_histograms_buckets: true
tag_by_endpoint: false
tags:
- image_id:********@sha256:57621adeb78e67c52e34ec1676d1ae898b252134838d60298c7446d0964551cc
- image_name:docker.io/istio/proxyv2
- image_tag:1.22.1
- kube_container_name:istio-proxy
- kube_deployment:reviews-v3
- kube_namespace:default
- kube_ownerref_kind:replicaset
- kube_qos:Burstable
- kube_replica_set:reviews-v3-7d99fd7978
- kube_service:reviews
- pod_phase:running
- short_image:proxyv2
...
```

These just represents the tags that automatically get added by the agent. But these are the ones where the agent scrapes the metrics from. 

Pod name would indeed be beneficial. The agent is able to also attach it but will need to set a param in the agent config for cardinality:
https://docs.datadoghq.com/containers/kubernetes/tag/?tab=datadogoperator#:~:text=pod_name,Pod%20metadata

Reply:
I see, very interesting! Do we have a common approach to kube based integrations on this matter?
Any downsides on adding the pod_name tag on the integration level?

Reply:
> Any downsides on adding the pod_name tag on the integration level?
I don't really believe so. If you need the tag for the integration to make more sense then it's understandable and will be a better usage than having it enabled agent wide for all integrations to pull in this tag. Since it's either all or nothing right now. The only concern is if the pod are short lived and will create new context every time it recreates. 

> Do we have a common approach to kube based integrations on this matter?
Right now, no. But I've encountered some concerns about these kube_* and pod_name tags collected by the agent natively. Essentially these are tagged from where these metrics are exposed (e.g. an exporter) but sometimes the metric themselves in prometheus has a label called namespace or pod name indicating where this specific metric is collected from by the exporter. Stuff like this can be remapped to their respective pod_name, kube_namespace for better alignment of the kubernetes integration/metrics with that of the ones reported by the exporter. But I'm not sure if this suits your case here. 

For example: 
```
kubevirt_vmi_cpu_system_usage_seconds_total{kubernetes_vmi_label_kubevirt_io_domain="testvm",kubernetes_vmi_label_kubevirt_io_nodeName="ip-192-168-58-28.eu-west-3.compute.internal",kubernetes_vmi_label_kubevirt_io_size="small",name="aws-kubevirt-vm-1",namespace="default",node="ip-192-168-58-28.eu-west-3.compute.internal"} 110.09
```

The origin of this metric I think is from a pod in the default namespace. If the agent is scraping this metric from the service that is exposing this is in namespace:foo for example, then all these metrics will come in with `kube_namespace:foo`, `namespace:default`. If you want to correlate this with like kubernetes.* metrics, then you can't align them using the kube_namespace tag.